### DISCUSSION TITLE
🚯 Spam prevention — a brief history and further ideas

### DISCUSSION THREAD

**Discussion by reidab** [calagator/calagator#701] [Ideas]
Throughout Calagator's history, we've tried several approaches to prevent spam events with varying levels of success. We're seeing a higher concentration of spammy events cropping up of late, so it's worth figuring out if there are new measures we should be taking. Spun off from https://github.com/calagator/calagator/issues/696 and https://github.com/calagator/calagator/pull/698, here's a brief history of some of what we've tried in the past and ideas I can remember. I'd be curious to hear other ideas or to dig in more on any of these.

### 🕳️ Trap Fields

Throwing an extra field in our submission forms that was hidden by CSS and would reject the event if any input was entered actually worked great back in 2009, but today's bots all run real web browsers and stuff. This is probably not providing us any protection anymore.

_There's not any modern version of this that makes sense, is there?_

### 🤖 Captchas

We've used ReCaptcha to guard forms since our trap fields stopped working. This did a reasonably good job at cutting down the number of things that got through, but its efficacy has slipped steadily over time. Now, with [readily availably APIs to defeat captchas](https://github.com/calagator/calagator/pull/698#issuecomment-1131199259) we're seeing even fewer things caught by these measures.

_Are there better captcha options we should be trying? 

### 🌐 IP Blocking

We have a number of blocked IPs and IP ranges that have been particularly egregious offenders in the past, but this list hasn't been updated in quite some time. We also have automated blocking for IPs that exceed certain rate or connection limits. 

_It may be worth examining our current influx to see if this could help. Do we live in an ipv6 world where this is useless yet?_

### 🙊 Word Blocking

We apply a manually-curated block list of words — err, regexes — to filter against spam. This also also is fairly effective, but hasn't been actively maintained in some time as the spam trends have shifted. 

That list is here for calagator.org https://github.com/calagator/calagator.org/blob/main/config/blacklist.txt

_Are there things we could add to this list that would stem our current wave of spam?_ 

### 🪴 Manual Gardening

Manual gardening is important for any wiki or wiki-adjacent calendar. We've always had people in the community keeping an eye on events as they were added and removing both bot-spam and random webinars from giant enterprise companies with no connection to Portland. In Calagator's most active days, this community effort usually caught anything that wasn't flagged by the other approaches fairly quickly and the effort was spread among enough people that it didn't feel particularly burdensome.

Currently, this approach faces the inherent challenges of pandemic life: there are (wisely) fewer legitimate events taking place, so spam comprises a higher percentage overall, and there are consequently fewer engaged users to spot and squash the spam.

### 👥 User Accounts + Email Verification

We don't have user accounts or any form of identity, but we could. At one of our first code sprints, we pushed back on the idea of user accounts being a requirement for every project and we've survived so far without them. There's a lot of powerful things that could be offered by accounts beyond just providing additional hedges against spam. While this is by no means a surefire solution, it's certainly a higher barrier than the current captchas and it could allow for easier gardening by allowing batch rollbacks of every action taken by a bad actor.

### 📚 Various Classifiers / ML Things

Various tools and APIs exist to attempt to classify content as spam or not spam. Much of the spam that we receive is the kind of thing these tools would be good at detecting. Some of these are paid services and some offer a free tier.

Examples include:
* https://cleantalk.org
* https://stopforumspam.com
* https://akismet.com
* https://www.oopspam.com

_Any other strong contenders?_

### 🤫 A semi-secret community phrase

[A recent proposal](https://github.com/calagator/calagator/pull/698) would require a password to add new events, seeded through  trusted people and shared readily to people known to be a part of the community sparked this discussion. I very much don't want  to require something like this without having exhausted other options. Even this seemingly small barrier to legitimate contributions would lead to a much smaller pool of potential editors and contributors.

---

* _Any other approaches we've missed and could consider?_
* _Are any of these existing approaches just not working at the moment due to a bug?_

**Comment by turoczy** [comment]
Appreciate all of this additional context, @reidab! Thanks for capturing it. Even though I've been a long-time Portland user and contributor, I always appreciate that there's a lot more going on behind the scenes than we see. So this is super helpful.

I just wanted to add that, in Portland at least, a good number of the new spam instances are written entirely in Cyrillic characters. I only see a few of those entries in the block list, so it's entirely possible that expanding the list of those particular words and phrases would help stifle some of this most recent activity.

**Comment by DGaffney** [comment]
Thank you so so much for the context - captchas still work, but they need to be non-standard to do it. Simple, stock implementations are very programmable these days. On the captcha route, one option would be to build something simple and bespoke and maybe related to calagator in some fun way (e.g. who is hosting an event tomorrow on the schedule). Would be a huge pain to program that for an attacker, and not too bad on our side of the equation. 

Another thing to consider - the "value" of the attack here from the attackers side is to embed reference links on Calagator that point back to their low-reputation sites. What if the URLs were simply not hyperlinks, and you didn't give them the value they were chasing? The systems are so automated that they likely might not notice, but this is attacking the exact incentive structure, at least.

**Comment by DGaffney** [comment]
Another thought here is should we put in a wait period before things get syndicated to RSS? At least that would stem the amount that these issues propagate to other communities...

**Reply by turoczy** [reply]
I believe there used to be a delay built in before things hit the RSS feed. I'm not sure if that delay is still present or not. It may just be a matter of folks not being as diligent on Calagator spam as we've been in the past. (So the delay isn't doing the same thing it used to.)

**Comment by DGaffney** [comment]
@reidab do you have a historical log of all deleted spam events? If so, if you sent that along to me along with legitimate events I'd be happy to play around with putting together a basic home-rolled classifier.

**Comment by DGaffney** [comment]
Another workaround @reidab - what if we didn't make accounts but at least enforced a click-in-email to confirm an event? That would at least raise costs for some of the most basic attackers out there (though of course, there are ways around that wall as well)

**Comment by jc00ke** [comment]
A few options I'd like to throw out:

* we train our own model with something like [bayespam](https://crates.io/crates/bayespam), though I'm guessing we don't have all the historical spam?
* we implement a basic email/username/TOTP authentication system. In fact, if we don't ever care to validate email address, I think there could be enough friction in setting up TOTP that spammers may leave us alone. I would hope that legit users would understand the predicament we're in.
* we implement a check that requires the person to at least use a calculator if not run some code. Multiplying or dividing some random numbers might be annoying enough for any human spam submitters and maybe we still rely on a captcha/TOTP for the bots.

This is a tricky, annoying problem. We can solve it though!

**Comment by jc00ke** [comment]
DEV/Forem is [battling this](https://dev.to/devteam/help-fight-spam-in-our-community-1ngb) as well. The [rel=ugc](https://moz.com/blog/nofollow-sponsored-ugc) URL param is an interesting approach, though it won't stop the submission which means it'll still populate the RSS feed.

**Comment by DGaffney** [comment]
Hello @reidab and @jc00ke! Been a minute. Independent of this thread, I've been developing an ML tool that leverages GPT models to rapidly design text-based machine learning classifiers (e.g. build quick and dirty MLs without tons of labeled data, tons of code, tons of time). I wanted to see how it did on this problem - I took a handful of 13 cases of spam/not spam from the PDX tech archives (there's a channel where spam gets booted to, and the good events stay put on the events channel), and let it train a ton of variations of the model - and I think I've got some that should work for the problem. Here's a temporary endpoint you can use to talk to the model and play with it a bit:

```ruby
require 'rest-client'
require 'json'

def get_spam_label(inference_model, api_key, text)
  url = "https://gpturk.cognitivesurpl.us/api/tasks/#{inference_model}/inferences"
  headers = { 'Content-Type' => 'application/json' }
  payload = { api_key: api_key, text: text }.to_json

  response = RestClient.post(url, payload, headers)
  JSON.parse(response.body)["label"]["parsed_label"].to_i
end

get_spam_label("is-this-a-spam-event--50a03ce4-9b0e-47d4-a8d9-0496dfde18b7", "59dd504c-ac1f-4ad7-9d73-73981c9cc8f9", text)
```

I think this should perform somewhere in the mid 90% range for accuracy on a balanced dataset of yes/no's, but we can play with it and tweak it more if you're interested in going this route! I'd be happy to provide this as an integration on an ongoing basis as a donation to Calagator if that's of interest.

**Reply by turoczy** [reply]
Thanks for sharing this @DGaffney! Those spammy Cyrillic posts are definitely detracting from the value of Calagator for folks like me who pull the RSS feed onto other platforms. It would be nice to get this resolved. Especially if it's an easy fix.

**Reply by DGaffney** [reply]
This ML approach is now in as a PR: https://github.com/calagator/calagator/pull/735

**Comment by jc00ke** [comment]
Expanding on user accounts and email verification, what if we used passkeys? I know they're new, but they could be a nice way to enforce that it's a real human entering the event details. We could pair them with a community passphrase as initial verification.

**Reply by DGaffney** [reply]
Hey @jc00ke! I don't know much about passkeys, but from a quick scan..... if you used them paired with a community passphrase, wouldn't the de facto community-passphrase-as-barrier-to-participation issue remain? You'd still potentially want anonymous, hitherto unknown individuals to be able to make events without access to the community prior to listing (or at least that was the (very reasonable) argument against a passphrase)?

**Reply by jc00ke** [reply]
I should have elaborated more: community passphrase for initial post, passkey to post after. I'm not sure how many new users we get on average.

**Comment by pnwmatt** [comment]
Restarting this conversation as today a post got through.

Sanitized and SFW: http://calagator.org/events/1250482475 . Here's the word frequency of the spam message:
```txt
the: 3
and: 2
women: 1
look: 1
confident: 1
comfortable: 1
which: 1
makes: 1
naked: 1
photos: 1
hit: 1
harder: 1
no: 1
awkward: 1
posing: 1
just: 1
sexy: 1
energy: 1
good: 1
framing: 1
sed "s/@//g" <<< htt@ps:@//b@lac@kpo@rnp@hot@os.c@om/      : 1
section: 1
delivers: 1
hot: 1
pics: 1
with: 1
real: 1
visual: 1
appeal: 1
```

The link was actually to 
sed "s/@//g" <<< htt@p:/@/c0@00f@fcc@2a1@.tr@ack@er.@ado@tmo@b.c@om/@pix@el/@vis@ite@?d=@500@0\\&r@=ht@tps@://@bla@ckp@orn@pho@tos@.co@m

I don't think the [blacklist.txt](https://github.com/calagator/calagator.org/blob/main/config/blacklist.txt) shouldn't be updated except maybe to add the ad tracker domain/maybe create a urlblacklist - but I don't think that these blacklists are a good long-term solution though.

My guess is the bot is trying "Post content -> Verify on homepage" (or retriggered by webcrawl -> search index has the spam domain) at some point.

Now that we're elbow deep in an AI world, I thought I'd update the conversation with notes from 2026:

- I don't think a community password is prudent.  I'm also a strong believer in open posting - but...
- I'd suggest passkey (just display name asked, no email verification) authentication.  All browsers and all password managers support this and budding new browsers are very likely to add support quickly
- Maybe a 72 hour modqueue for users or anons who don't have a satisfactory posting history?
- All users are mods (for now)
- Can't mod your own post, and if/when it becomes an issue, we just switch to new users not being mods
- Mod can subscribe to the modqueue if they want in the same way they subscribe to the current feed
- Authenticated Users eventually graduate to not being in the modqueue.  If an account starts getting weird, then we adjust as needed.

For events IRL we could have a QR code that promotes accounts to something like "human verified" which promotes the account.

I'd be happy to contribute to a design doc/implementation plan as a pitch for collaboration once I get some feedback from the community.

**Reply by jc00ke** [reply]
I like the passkey idea. We could also look into a system like https://github.com/mitchellh/vouch though I see it hasn't been standardized yet.
